lsbfdump/src/main/scala/org/apache/daffodil/lsbfDump/LSBFDump.scala

```diff
@@ -0,0 +1,188 @@
+
+import java.io.IOException
+import java.nio.ByteBuffer
+import java.nio.channels.FileChannel
+import java.nio.file.NoSuchFileException
+import java.nio.file.Paths
+
+object LSBFDump {
+
+  private val helpText: String =
+    """Usage: lsbfdump --file <filename> [--offset <offset>] [--length <numBytes>] [--noAddress] [--help]
+      |
+      |<filename>   : The file to read bytes from or '-' for standard input.
+      |[offset]     : The starting offset in the file (default is 0).
+      |[length]     : The number of bytes to display (default is 128).
+      |--noAddress  : Do not display the address of each byte line.
+      |--help       : Display this help information.
+      |
+      |Examples:
+      | Default usage (128 bytes from standard input, starting at offset 0, with addresses):
+      |   lsbfdump --file -
+      |
+      | With specific file, offset and byte count:
+      |   lsbfdump --file filename --offset 10 --length 64
+      |
+      | With --noAddress to hide addresses:
+      |   lsbfdump --file filename --offset 10 --length 64 --noAddress
+      |""".stripMargin
+
+  private def usageError(): Unit = {
+    System.err.println("Error: Invalid arguments provided.")
+    System.err.println(helpText)
+    sys.exit(1)
+  }
+
+  def main(args: Array[String]): Unit = {
+    val parsedArgs = parseArgs(args)
+
+    if (parsedArgs.contains("help")) {
+      println(helpText)
+      sys.exit(0)
+    }
+
+    if (parsedArgs.contains("invalid")) usageError()
+
+    val filename = parsedArgs.getOrElse("file", "-")
+    val offset = parsedArgs.get("offset").map(_.toInt).getOrElse(0)
+    val numBytes = parsedArgs.get("length").map(_.toInt).getOrElse(128)
+    val showAddress = !parsedArgs.contains("noAddress")
+
+    val lines = lsbfDumpFile(filename, offset, numBytes, showAddress)
+    lines.foreach(println(_))
+  }
+
+  def lsbfDumpFile(
+    filename: String,
+    offset: Int,
+    numBytes: Int,
+    showAddress: Boolean,
+  ): Seq[String] = {
+    val bytes =
+      try {
+        if (filename == "-") {
+          readStdInBytes(offset, numBytes)
+        } else {
+          readFileBytes(filename, offset, numBytes)
+        }
+      } catch {
+        case e: NoSuchFileException =>
+          System.err.println(s"Error: File not found - ${e.getMessage}")
+          sys.exit(1)
+        case e: IOException =>
+          System.err.println(s"Error: An I/O error occurred - ${e.getMessage}")
+          sys.exit(1)
+      }
+    val lines = lsbfDump(bytes, offset, showAddress)
+    lines
+  }
+
+  def lsbfDump(
+    bytes: Seq[Byte],
+    offset: Int,
+    showAddress: Boolean,
+  ): Seq[String] = {
+
+    val lines = bytes
+      .grouped(4)
+      .zipWithIndex
+      .map { case (group, index) =>
+        val byteStrings = group.reverse.map(byteToBinaryString).mkString(" ")
+        val addressString = if (showAddress) f" | 0x${index * 4 + offset}%08X" else ""
+        byteStrings + addressString
+      }
+      .toList
```

Comment:
Mentioned above, but reading all the bytes an mapping them feels like a lot of memory usage for something that can probably be streamed fairly easily.